File: step_15_graphs/15.2_bfs_dfs/IB131_possibility_of_finishing_all_courses/solution.py
"""
Problem: Possibility of Finishing All Courses
Difficulty: MEDIUM | XP: 25
Source: InterviewBit

Given N courses and prerequisites [a, b] meaning "must take b before a",
determine if all courses can be finished (detect cycle in directed graph).
"""

from typing import List
from collections import deque


# ============================================================
# APPROACH 1: BRUTE FORCE
# DFS 3-color cycle detection (WHITE=0, GRAY=1, BLACK=2)
# Time: O(V + E)  |  Space: O(V + E)
# ============================================================
def brute_force(num_courses: int, prerequisites: List[List[int]]) -> bool:
    adj = [[] for _ in range(num_courses)]
    for a, b in prerequisites:
        adj[b].append(a)  # b must come before a

    color = [0] * num_courses  # 0=white, 1=gray, 2=black

    def has_cycle(node):
        color[node] = 1
        for neighbor in adj[node]:
            if color[neighbor] == 1:
                return True
            if color[neighbor] == 0 and has_cycle(neighbor):
                return True
        color[node] = 2
        return False

    for i in range(num_courses):
        if color[i] == 0 and has_cycle(i):
            return False
    return True


# ============================================================
# APPROACH 2: OPTIMAL
# Kahn's Algorithm (BFS Topological Sort)
# Count processed nodes — if == numCourses, no cycle
# Time: O(V + E)  |  Space: O(V + E)
# ============================================================
def optimal(num_courses: int, prerequisites: List[List[int]]) -> bool:
    adj = [[] for _ in range(num_courses)]
    in_degree = [0] * num_courses

    for a, b in prerequisites:
        adj[b].append(a)
        in_degree[a] += 1

    queue = deque(i for i in range(num_courses) if in_degree[i] == 0)
    processed = 0

    while queue:
        node = queue.popleft()
        processed += 1
        for neighbor in adj[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return processed == num_courses


# ============================================================
# APPROACH 3: BEST
# Iterative DFS using explicit stack — no recursion limit issues
# Time: O(V + E)  |  Space: O(V + E)
# ============================================================
def best(num_courses: int, prerequisites: List[List[int]]) -> bool:
    adj = [[] for _ in range(num_courses)]
    for a, b in prerequisites:
        adj[b].append(a)

    state = [0] * num_courses  # 0=unvisited, 1=in-stack, 2=done

    for start in range(num_courses):
        if state[start] != 0:
            continue
        # stack stores (node, neighbor_index)
        stack = [(start, 0)]
        state[start] = 1

        while stack:
            node, idx = stack[-1]
            neighbors = adj[node]
            if idx < len(neighbors):
                stack[-1] = (node, idx + 1)
                nxt = neighbors[idx]
                if state[nxt] == 1:
                    return False  # cycle detected
                if state[nxt] == 0:
                    state[nxt] = 1
                    stack.append((nxt, 0))
            else:
                state[node] = 2
                stack.pop()

    return True


if __name__ == "__main__":
    print("=== Possibility of Finishing All Courses ===")

    print(f"BruteForce [2, [[1,0]]]:     {brute_force(2, [[1,0]])}")         # True
    print(f"Optimal    [2, [[1,0]]]:     {optimal(2, [[1,0]])}")              # True
    print(f"Best       [2, [[1,0]]]:     {best(2, [[1,0]])}")                 # True

    print(f"BruteForce cycle:            {brute_force(2, [[1,0],[0,1]])}")    # False
    print(f"Optimal    cycle:            {optimal(2, [[1,0],[0,1]])}")         # False
    print(f"Best       cycle:            {best(2, [[1,0],[0,1]])}")            # False

    print(f"4 courses no cycle: {optimal(4, [[1,0],[2,0],[3,1],[3,2]])}")      # True

File: step_15_graphs/15.2_bfs_dfs/IB131_possibility_of_finishing_all_courses/Solution.java
import java.util.*;

/**
 * Problem: Possibility of Finishing All Courses
 * Difficulty: MEDIUM | XP: 25
 * Source: InterviewBit
 *
 * Given N courses and prerequisites as edges [a, b] meaning "must take b before a",
 * determine if all courses can be finished (i.e., no cycle in directed graph).
 *
 * @author DSA_Nova
 */

// ============================================================
// APPROACH 1: BRUTE FORCE
// DFS-based cycle detection using 3-color marking (WHITE=0, GRAY=1, BLACK=2)
// Time: O(V + E)  |  Space: O(V + E)
// ============================================================
class BruteForce {
    static int[] color;
    static List<List<Integer>> adj;

    static boolean hasCycle(int node) {
        color[node] = 1; // GRAY = being processed
        for (int neighbor : adj.get(node)) {
            if (color[neighbor] == 1) return true; // back edge => cycle
            if (color[neighbor] == 0 && hasCycle(neighbor)) return true;
        }
        color[node] = 2; // BLACK = done
        return false;
    }

    static boolean canFinish(int numCourses, int[][] prerequisites) {
        adj = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) adj.add(new ArrayList<>());
        for (int[] pre : prerequisites) adj.get(pre[1]).add(pre[0]);

        color = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            if (color[i] == 0 && hasCycle(i)) return false;
        }
        return true;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL
// Kahn's Algorithm (BFS Topological Sort) — count processed nodes
// If all nodes processed, no cycle exists.
// Time: O(V + E)  |  Space: O(V + E)
// ============================================================
class Optimal {
    static boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adj = new ArrayList<>();
        int[] inDegree = new int[numCourses];

        for (int i = 0; i < numCourses; i++) adj.add(new ArrayList<>());
        for (int[] pre : prerequisites) {
            adj.get(pre[1]).add(pre[0]);
            inDegree[pre[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) queue.offer(i);
        }

        int processed = 0;
        while (!queue.isEmpty()) {
            int node = queue.poll();
            processed++;
            for (int neighbor : adj.get(node)) {
                if (--inDegree[neighbor] == 0) queue.offer(neighbor);
            }
        }
        return processed == numCourses;
    }
}

// ============================================================
// APPROACH 3: BEST
// Iterative DFS using explicit stack — avoids recursion limit issues
// Time: O(V + E)  |  Space: O(V + E)
// ============================================================
class Best {
    static boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) adj.add(new ArrayList<>());
        for (int[] pre : prerequisites) adj.get(pre[1]).add(pre[0]);

        // 0=unvisited, 1=in-stack, 2=done
        int[] state = new int[numCourses];

        for (int start = 0; start < numCourses; start++) {
            if (state[start] != 0) continue;
            Deque<int[]> stack = new ArrayDeque<>(); // [node, neighborIndex]
            stack.push(new int[]{start, 0});
            state[start] = 1;

            while (!stack.isEmpty()) {
                int[] top = stack.peek();
                int node = top[0];
                List<Integer> neighbors = adj.get(node);

                if (top[1] < neighbors.size()) {
                    int next = neighbors.get(top[1]++);
                    if (state[next] == 1) return false; // cycle
                    if (state[next] == 0) {
                        state[next] = 1;
                        stack.push(new int[]{next, 0});
                    }
                } else {
                    state[node] = 2;
                    stack.pop();
                }
            }
        }
        return true;
    }
}

public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Possibility of Finishing All Courses ===");

        // 2 courses, no cycle: can finish
        System.out.println("BruteForce [2, [[1,0]]]:     " + BruteForce.canFinish(2, new int[][]{{1, 0}}));   // true
        System.out.println("Optimal    [2, [[1,0]]]:     " + Optimal.canFinish(2, new int[][]{{1, 0}}));      // true
        System.out.println("Best       [2, [[1,0]]]:     " + Best.canFinish(2, new int[][]{{1, 0}}));         // true

        // 2 courses, cycle: cannot finish
        System.out.println("BruteForce cycle [2,0],[0,1]: " + BruteForce.canFinish(2, new int[][]{{1, 0}, {0, 1}})); // false
        System.out.println("Optimal    cycle:             " + Optimal.canFinish(2, new int[][]{{1, 0}, {0, 1}}));    // false

        // 4 courses with complex deps
        System.out.println("4 courses no cycle: " + Optimal.canFinish(4, new int[][]{{1,0},{2,0},{3,1},{3,2}})); // true
    }
}
